Support EXT_mesh_primitive_restart glTF extension

eringram opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

<!-- Describe your changes in detail -->

<!-- Consider: Why is this change required? What problem does it solve? -->

<!-- Include screenshots if appropriate -->

This PR adds support for the EXT_mesh_primitive_restart glTF extension. It is based on the implementation in iTwin.js-- see `getMeshPrimitives` in [GltfReader.ts](https://github.com/iTwin/itwinjs-core/blob/17c6f833ed1753f4e969e0edaab8aaf315ab23b8/core/frontend/src/tile/GltfReader.ts#L436). 

"Primitive restart" refers to starting a new primitive of the same type as the current primitive when a specific index value is encountered (the maximum value for that index buffer type). It is useful for batching multiple primitives of the same type in a single draw call. See the README in [this PR](https://github.com/KhronosGroup/glTF/pull/2478) for more detailed explanation and a simple example.

## Issue number and link

<!-- If it fixes an open issue, link to the issue here -->

<!-- Consider: If suggesting a new feature or change, discuss it in an issue first. -->

Addresses https://github.com/CesiumGS/cesium/issues/12764

## Testing plan

<!-- Describe in detail how you tested your changes. If this fixes a bug, list the steps to reproduce the original issue. -->

- Added a unit test to GltfLoaderSpec.js that uses the MeshPrimitiveRestart.gltf test model and confirms that the new `getMeshPrimitives` function reduces the length of the primitives array from 8 to 4 (effectively combining the 4 groups of 2 primitives each).
    - The test model includes primitive restart groups for all 4 supported modes.
- Added unit tests that confirm the extension is **not** used if the glTF model violates the spec (i.e. uses a primitive in multiple groups, a primitive has an unsupported mode, etc.)
- You can also manually test by opening MeshPrimitiveRestart.gltf in Sandcastle and stepping through the code to confirm the extension is used. This is what the model looks like:

<img width="670" height="498" alt="image" src="https://github.com/user-attachments/assets/e0bdc60b-f011-471e-bb5e-1b0bf6969300" />

(the 4 groups are two green triangles, two blue triangles, two green line strips, and two blue line loops)

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

ggetz commented:
@eringram Is this PR ready for full or partial review? If so, could you please update the PR description with the current status? Thanks!

eringram commented:
> @eringram Is this PR ready for full or partial review? If so, could you please update the PR description with the current status? Thanks!

Sorry just updated the description! The code in GltfLoader.js is ready for review and still working on more thorough unit tests.

javagl commented on packages/engine/Source/Scene/GltfLoader.js:
```diff
@@ -2394,6 +2395,75 @@ function loadInstanceFeaturesLegacy(
+  // Start with a copy of mesh.primitives. For each group, replace the first primitive in the group with a primitive representing the entire group,
+  // and set the rest of the primitives in the group to `undefined`.
+  // This allows us to identify which remaining primitives do not use primitive restart, and any errors involving a primitive appearing in more than one group.
+  const primitives = [...meshPrimitives];
```
We had some issues with `push.apply` because it puts all elements on the stack (yeah, sounds crazy, but that's the way it is), and the same applies to the `[...spread]` operator (see https://github.com/CesiumGS/cesium/issues/12053 ). I don't know if we want to be prepared for meshes with 150000 primitives...

markschlosseratbentley replied:
Thanks for the `addAllToArray` suggestion, @javagl.  @eringram, once we add usage of that function to this PR, we should consider doing something similar to this parallel extension code in iTwin.js if necessary.

eringram replied:
Thanks, just updated this to use `addAllToArray`.

eringram commented:
I updated the MeshPrimitiveRestart.gltf test data to include primitive groups for all the valid modes for the extension (line loop, line strip, triangle strip, triangle fan). Also added more unit tests to ensure `getMeshPrimitives` falls back to the default `mesh.primitives` if EXT_mesh_primitive_restart violates the spec. Two of those tests are failing, where the extension is used even if incorrect, so still fixing that.

Also, @javagl @ggetz any idea why [this test](https://github.com/CesiumGS/cesium/pull/12792/files#diff-8f11fb72d71b01b789e5660c80ad63658fa8768871b77786e95b636aa918c6c4R4279) that calls `toBeRejectedWithError` is failing to match the error message regex?

This is the output when it runs and says it fails, but the errors look identical to me:

```
3) does not load with EXT_mesh_primitive_restart if a group's indices accessor is invalid
     Scene/GltfLoader
     Expected a promise to be rejected with RuntimeError: /Failed to load glTF
Failed to load index buffer
indexDataType is required and must be a valid IndexDatatype constant./ but it was rejected with RuntimeError: Failed to load glTF
Failed to load index buffer
indexDatatype is required and must be a valid IndexDatatype constant.
Original stack:
Original stack:
Error
    at new DeveloperError (packages/engine/Source/Core/DeveloperError.js:39:11 <- Build/CesiumUnminified/Cesium.js:8893:13)
    at IndexDatatype.getSizeInBytes (packages/engine/Source/Core/IndexDatatype.js:62:9 <- Build/CesiumUnminified/Cesium.js:26912:11)
    at createIndicesTypedArray (packages/engine/Source/Scene/GltfIndexBufferLoader.js:248:35 <- Build/CesiumUnminified/Cesium.js:70949:45)
    at loadFromBufferView2 (packages/engine/Source/Scene/GltfIndexBufferLoader.js:227:37 <- Build/CesiumUnminified/Cesium.js:70930:39)
    at async Promise.all (index 7)
Handler stack:
Error
    at new RuntimeError (packages/engine/Source/Core/RuntimeError.js:38:11 <- Build/CesiumUnminified/Cesium.js:13822:13)
    at ResourceLoader.getError (Build/Ce ... ....
    at <Jasmine>
    at UserContext.<anonymous> (packages/engine/Specs/Scene/GltfLoaderSpec.js:4263:9 <- Build/Specs/SpecList.js:209938:9)
    at <Jasmine>
```

eringram commented:
This is ready for full review.

Anyone have an idea why the `prettier` format check is failing? It does not cause a warning for CHANGES.md when I run locally.

javagl commented:
(Quick, while in another call:) 

Regarding the formatting: Indeed, something is strange here: https://github.com/CesiumGS/cesium/pull/12801#issuecomment-3161155566

The test that is failing _now_ is related to some error message not matching the expected one, at https://github.com/CesiumGS/cesium/pull/12792/files#diff-8f11fb72d71b01b789e5660c80ad63658fa8768871b77786e95b636aa918c6c4R4277 . 

A bit more generally, I think that it's a bit dubious to check for the _exact_ error message, even though this is done in many places. I think that this "overspecifies" the test. (I hope there are no plans on internationalization, FWIW...). One could try to "justify" that, but only with the severe lack of differentiation between _types_ of errors: There should be far more than just `RuntimeError` and `DeveloperError`, and the latter is used in many places where the _developer_ actually has no influence on anything, and it should rather be a `DataError` or so. 

However, the test here _could_ just be changed to expect a type error, but ... of course, such an error should preferably not be thrown to the user either...

eringram commented:
> Regarding the formatting: Indeed, something is strange here: [#12801 (comment)](https://github.com/CesiumGS/cesium/pull/12801#issuecomment-3161155566)

Ok good to know it wasn't just me, I ended up just reverting the whitespace. Not sure how to fix the discrepancy it but it seems related to the `blanks-around-fences` markdownlint rule.

> A bit more generally, I think that it's a bit dubious to check for the _exact_ error message, even though this is done in many places. I think that this "overspecifies" the test. (I hope there are no plans on internationalization, FWIW...). One could try to "justify" that, but only with the severe lack of differentiation between _types_ of errors: There should be far more than just `RuntimeError` and `DeveloperError`, and the latter is used in many places where the _developer_ actually has no influence on anything, and it should rather be a `DataError` or so.
> 
> However, the test here _could_ just be changed to expect a type error, but ... of course, such an error should preferably not be thrown to the user either...

These considerations make sense, you're right that the lack of error types kind of leads to comparing the error messages like this. I tried to use a generic `Error` type, but now see that the test is timing out. Any idea about that? I see other tests using `await expectAsync(...).toBeRejectedWithError()` that don't appear to increase their timeout past 5000ms.

eringram commented on packages/engine/Source/Scene/GltfLoader.js:
```diff
@@ -2394,6 +2395,76 @@ function loadInstanceFeaturesLegacy(
+ * @returns {object[]} An array of mesh primitives
+ * @private
+ */
+function getMeshPrimitives(mesh) {
```
I agree, moved to a new file.

eringram commented:
@pmconne Not required, but checking to see if you intended to review this
